### PR TITLE
(don't review) Added ADC+DMA+TIMER

### PR DESCRIPTION
**Pull request by markomarkusi**
### Summary
Minimal setup for using an ADC with DMA plus a  1kHz Timer to trigger the ADC conversion sequence

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: boards/PDM/Src/main.c [hunk 1]
```diff
@@ -183,6 +194,9 @@ int main(void)
 
     can_rx = App_CanRx_Create();
 
+    HAL_ADC_Start_DMA(&hadc1, (uint32_t*)adc_readings, NUM_ADC_CHANNELS);
+    HAL_TIM_Base_Start(&htim2);
```
Comment (markomarkusi) [inline boards/PDM/Src/main.c:198]: 1kHz timer used for triggering ADC conversion sequence

File: boards/shared/Src/Io/Io_Adc.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *hadc)
+{
+    // If you were to over-sample this is where you can perform:
+    // Filter_ADC_Readings(ADC_CHANNEL_COUNT, ADC_FILTER_SIZE,
+    // Unfiltered_ADC_Readings, Filtered_ADC_Readings);
+
+    v1 = Convert_ADC_Voltage(adc_readings[CHANNEL_2]);
+    v2 = Convert_ADC_Voltage(adc_readings[CHANNEL_7]);
+    v3 = Convert_ADC_Voltage(adc_readings[CHANNEL_8]);
+    v4 = Convert_ADC_Voltage(adc_readings[CHANNEL_9]);
```
Comment (markomarkusi) [inline boards/shared/Src/Io/Io_Adc.c:15]: For each voltage/current measurement you would want to implement a function in _IoAdc.c_ that would look something like:
```
float Get_Vbat(uint16_t adc_value,)
{
	float voltage = Convert_ADC_Voltage(adc_value);
	return voltage*VBAT_VOLTAGE_GAIN_RATIO;
}

```

**Comment by thekenu** [comment]
This PR is for a meeting we have tomorrow, no need to review for now

File: boards/shared/Inc/Io/Io_SharedFilters.h [hunk 3]
```diff
@@ -23,3 +23,13 @@ void Io_SharedFilters_LowPassFilter(
...
+
+/**
+  * @brief  Filters ADC readings by computing the mean of the previous ADC_FILTER_SIZE readings (oversampling filter)
+  * @param  ChannelCount Number of ADC channels used
+  * @param  FilterSize Oversampling filter size (number of readings)
+  * @param  ADC_Readings Pointer to array containing the unfiltered ADC readings
+  * @param  Filtered_ADC_Readings Pointer to array in which the filtered ADC readings will be stored
+  * @retval None
+ */
+void Filter_ADC_Readings(uint32_t ChannelCount, uint32_t FilterSize, uint16_t *ADC_Readings, float *Filtered_ADC_Readings);
```
Comment (markomarkusi) [inline boards/shared/Inc/Io/Io_SharedFilters.h:35]: @thekenu I believe you created this function, but I can't find it on consolidated-firmware #stolencode

File: boards/PDM/PDM.ioc [hunk 4]
```diff
@@ -1,37 +1,44 @@
...
+ADC1.OffsetNumber-9\#ChannelRegularConversion=ADC_OFFSET_NONE
 ADC1.Overrun=ADC_OVR_DATA_OVERWRITTEN
+ADC1.Rank-10\#ChannelRegularConversion=4
 ADC1.Rank-7\#ChannelRegularConversion=1
+ADC1.Rank-8\#ChannelRegularConversion=2
+ADC1.Rank-9\#ChannelRegularConversion=3
 ADC1.Resolution=ADC_RESOLUTION_12B
-ADC1.SamplingTime-7\#ChannelRegularConversion=ADC_SAMPLETIME_1CYCLE_5
-ADC1.ScanConvMode=ADC_SCAN_DISABLE
-ADC1.SubFamily=STM32F302xC
```
Comment (pasitlaothamatas) [inline boards/PDM/PDM.ioc:34]: Now that we are using 1kHz interrupts to initiate the conversion we can probs get away with less sample time cycles

image

Comment (markomarkusi) [reply boards/PDM/PDM.ioc:34]: Sampling rate is dependent on the [external impedance ](https://www.st.com/resource/en/application_note/cd00211314-how-to-get-the-best-adc-accuracy-in-stm32-microcontrollers-stmicroelectronics.pdf#page=19)seen by the adc, the datasheet has a section on calculating Rain. You have to calculate your min sampling rate and then you can set it to that.

**Comment by thekenu** [comment]
No longer needed after offline discussion

**Pull request closed without merging**